README
== Purpose ==

episoder is a tool to tell you about new episodes of your favourite TV shows.


== Installation ==

* Run 'easy_install episoder-0.7.2.tar.gz'
* Create a cron job to have your database rebuilt once a day
* Add "episoder" to your ~/.bashrc, ~/.bash_profile, /etc/bash.bashrc or /etc/profile if you want to see your upcoming shows every time you start a shell. Consult the bash docs to find out which file you want to use.


== Configuration ==

See "man episoder"


== Contact ==

If you find any bugs or have ideas on how to improve episoder, please use the
issue tracker on our website at https://github.com/cockroach/episoder


CHANGELOG
== v0.8.3 (27 Sep 2017) ==
 * Fixed a crash related to unexpected null values in TVDB JSON data
 * Use correct path for example config file in setup.py (fixes #7)

== v0.8.2 (7 Jul 2017) ==
 * Fixed a crash when rendering episodes without a title

== v0.8.1 (24 Jun 2017) ==
 * Handle failures to load TVDB API pages

== v0.8.0 (7 Jun 2017) ==
 * Switched from distutils to setuptools
 * We now use the new TVDB JSON API
 * Rewrote the epguides.com parser in pure Python
 * Rewrote large parts of the database code
 * Episoder should now work fully with Python 3
 * The update command can now be limited to a fixed number of shows
 * Updated man page
 * Thanks to Chris Spencer (chrisspen) for his contributions
 * Added option to only list active shows (chrisspen)
 * Added the number of episodes to the list of shows (chrisspen)
 * Added email notification feature via 'episoder notify' (chrisspen)

== v0.7.3 (7 Nov 2015) ==
 * Fixed a syntax error in the list of shows
 * Fixed season number parsing for epguides.com (thanks to jethrogb)

== v0.7.2 (26 Sep 2015) ==
 * Dropped unneeded dependency on BeautifulSoup
 * Fixed episode number parsing for epguides.com (thanks to jethrogb)
 * Improved python3 compatibility

== v0.7.1 (27 Apr 2014) ==
 * Accept unicode characters in prodnum fields

== v0.7.0 (14 Oct 2013) ==
 * Added support for TheTVDB.com
 * Removed support for tv.com
 * Changed command-line arguments. See the man page for details
 * Removed the ability to specify shows in the configuration file
 * Reorganized the command line options. Check the documentation for details
 * Updated the epguides.com parser to ignore trailer links in episode names
 * Fixed some more encoding issues in the console renderer
 * Changed license to GPLv3

== v0.6.5 (6 Oct 2010) ==
 * Updated the epguides.com parser to ignore links to trailers in episode names

== v0.6.4 (25 Jun 2010) ==
 * Filter javascript from tv.com pages because it confuses beautifulsoup 3.1

== v0.6.3 (23 Jun 2010) ==
 * The epguides parser now properly reads the season number for more shows

== v0.6.2 (08 Nov 2009) ==
 * Updated tv.com parser (show names need to be lower-case now!)
 * Added -S option to specify a show id for (forced) updates

== v0.6.1 (26 Aug 2009) ==
 * Episoder can now talk to any database supported by sqlalchemy (see README)
 * Tell awk to process utf8 (for epguides parser)
 * Catch (and properly log) some more HTTP exceptions

== v0.6.0 (8 Aug 2009) ==
 * Episoder now keeps a persistent database of shows and episodes
 * Added new options (-L, -a, -r, -E, -D) to manage the database
 * The parsers now know whether a show is still running or not
   * Active (currently running) shows are updated every two days
   * Inactive (ended) shows are updated every two weeks
   * The new -B command overrides those intervals and forces an update
 * Show names can no longer be overridden in the configuration file
 * Don't report an error when no sources are defined in the config file
 * The tv.com parser now also accepts urls containing '!', '+' or '%'
 * The -d option now also accepts the number of days instead of a fixed date
 * Documentation updated

== v0.5.3 (26 Jul 2009) ==
 * Added support for another epguides formatting style (used eg. for "Eureka")
 * Added option to suppress colors in console output
 * Added a filter to remove invalid characters from epguides content
 * Fixed some other parsing issues in the epguides parser
 * Improved compatibility with older python versions

== v0.5.2 (30 Jun 2009) ==
 * Reworked parts of the code to improve compatibility with older libraries
 * Fixed some encoding issues in the console renderer

== v0.5.1 (23 Jun 2009) ==
 * Thanks to Philip Douglass for his bug reports and patches
 * The epguides parser now properly closes all file handles and works with mawk
 * The tv.com parser would crash on missing production/season/episode numbers
 * The tv.com parser now accepts urls that contain a dash
 * Switched to using sqlalchemy to access the database
 * Properly implemented the user-agent configuration option
 * Removed old FAQ
 * Fixed some text encoding issues
 * Small fixes to the man page

== v0.5.0 (18 May 2009) ==
 * Rewrote the whole thing in python
 * Wrote a brand new tv.com parser
 * Small fixes in the epguides.com parser
 * Parsed data is now stored in an sqlite3 database

== v0.4.7 (24 Feb 2007) ==
 * Fixed a parsing issue introduced with 0.4.6

== v0.4.6 (24 Feb 2007) ==
 * Discovered yet another formatting possibility on epguides.com (used for
   instance for the show "Midsomer Murders") and adapted the parser accordingly
 * Slightly updated the man page

== v0.4.5 (23 Jan 2007) ==
 * Fixed some epguides parser issues

== v0.4.4 (19 Nov 2006) ==
 * Fixed the -i option when building the database (did the exact opposite of
   what it was supposed to do)

== v0.4.3 (24 Oct 2006) ==
 * We now actually honor the -i option when building the database
 * Added the possibility to override a show's name in the config file. This was needed for the new webisoder, a web-based episoder. Find out more about it at http://www.webisoder.net/
 * Oops, forgot to update episoder's version number in the last release...
 * Fixed a bug in the new epguides.com parser which would not drop episodes without a date

== v0.4.2 (21 Oct 2006) ==
 * Don't ignore the config file passed with -c anymore when building the db
 * Added parser for epguides.com (closes #38)

== v0.4.1 (20 Apr 2006) ==
 * Did some minor internal changes to get episoder to work on non-linux systems
 * Removed long options
 * Hotfix for the tv.com website changes

== v0.4.0 (3 Dec 2005) ==
 * Re-organized options: -vv is now -w, -rc is now -c (or --rc), -nd is gone (use --nodate), -sa is gone (use -s and --nodate instead), -i is gone (see below)
 * Search is now alway case-insensitive
 * If no action is specified, episoder now defaults to showlist
 * Removed the tvtome dummy plugin

== v0.3.0 (9 Aug 2005) ==

pyepisoder/sources.py
# episoder, https://github.com/cockroach/episoder
#
# Copyright (C) 2004-2017 Stefan Ott. All rights reserved.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import logging

from re import search, match
from json import dumps
from datetime import datetime

import requests

from .database import Episode, Show


class ParserSelector(object):

	class _Inner(object):

		def __init__(self):

			self._tvdb = TVDB()
			self._epguides = Epguides()
			self._tvcom = TVCom()

		def parser_for(self, url):

			for parser in [self._tvdb, self._epguides, self._tvcom]:
				if parser.accept(url):
					return parser

			return None

	_instance = None

	@staticmethod
	def instance():

		if not ParserSelector._instance:
			ParserSelector._instance = ParserSelector._Inner()

		return ParserSelector._instance


class InvalidLoginError(Exception):

	pass


class TVDBShowNotFoundError(Exception):

	pass


class TVDBNotLoggedInError(Exception):

	pass


class TVDBOffline(object):

	def __init__(self, tvdb):

		self._tvdb = tvdb
		self._log = logging.getLogger("TVDB (offline)")

	def __str__(self):

		return "thetvdb.com parser (ready)"

	def __repr__(self):

		return "<TVDBOffline>"

	def _post_login(self, data, user_agent):

		url = "https://api.thetvdb.com/login"
		head = {"Content-type": "application/json",
			"User-Agent": user_agent}
		body = dumps(data).encode("utf8")
		response = requests.post(url, body, headers=head)
		data = response.json()

		if response.status_code == 401:
			raise InvalidLoginError(data.get("Error"))

		self._log.info("Successful login")
		return data.get("token")

	def lookup(self, _, __):

		raise TVDBNotLoggedInError()

	def login(self, args):

		body = {"apikey": args.tvdb_key}
		self.token = self._post_login(body, args.agent)

	def parse(self, _, __, ___):

		raise TVDBNotLoggedInError()

	def _set_token(self, token):

		self._tvdb.change(TVDBOnline(token))

	token = property(None, _set_token)


class TVDBOnline(object):

	def __init__(self, token):

		self._token = token
		self._log = logging.getLogger("TVDB (online)")

	def __str__(self):

		return "thetvdb.com parser (authorized)"

	def __repr__(self):

		return "<TVDBOnline>"

	def _get(self, url, params, agent):

		url = "https://api.thetvdb.com/%s" % url
		head = {"Content-type": "application/json",
			"User-Agent": agent,
			"Authorization": "Bearer %s" % self._token}
		response = requests.get(url, headers=head, params=params)
		data = response.json()

		if response.status_code == 404:
			raise TVDBShowNotFoundError(data.get("Error"))

		return data

	def _get_episodes(self, show, page, agent):

		show_id = int(show.url)
		opts = {"page": page}
		result = self._get("series/%d/episodes" % show_id, opts, agent)
		return (result.get("data"), result.get("links"))

	def lookup(self, term, agent):

		def mkshow(entry):

			name = entry.get("seriesName")
			url = str(entry.get("id")).encode("utf8").decode("utf8")
			return Show(name, url=url)

		matches = self._get("search/series", {"name": term}, agent)
		return map(mkshow, matches.get("data"))

	def login(self, args):

		pass

	def _fetch_episodes(self, show, page, agent):

		def mkepisode(row):

			num = int(row.get("airedEpisodeNumber", "0") or 0)
			aired = row.get("firstAired")
			name = row.get("episodeName") or u"Unnamed episode"
			season = int(row.get("airedSeason", "0") or 0)
			aired = datetime.strptime(aired, "%Y-%m-%d").date()
			pnum = u"UNK"

			self._log.debug("Found episode %s", name)
			return Episode(name, season, num, aired, pnum, 0)

		def isvalid(row):

			return row.get("firstAired") not in [None, ""]

		(data, links) = self._get_episodes(show, page, agent)
		valid = filter(isvalid, data)
		episodes = [mkepisode(row) for row in valid]

		# handle pagination
		next_ = links.get("next") or 0
		if next_ > page:
			try:
				more = self._fetch_episodes(show, next_, agent)
				episodes.extend(more)
			except TVDBShowNotFoundError:
				msg = "Error parsing %s: failed to load page %d"
				self._log.error(msg, show.name, next_)

		return episodes

	def parse(self, show, db, user_agent):

		result = self._get("series/%d" % int(show.url), {}, user_agent)
		data = result.get("data")

		# update show data
		show.name = data.get("seriesName", show.name)
		show.updated = datetime.now()

		self._log.debug("Updating show '%s'", show.name)

		if data.get("status") == "Continuing":
			show.status = Show.RUNNING
		else:
			show.status = Show.ENDED

		# load episodes
		episodes = sorted(self._fetch_episodes(show, 1, user_agent))
		for (idx, episode) in enumerate(episodes):

			episode.totalnum = idx + 1
			db.add_episode(episode, show)

		db.commit()


class TVDB(object):

	def __init__(self):

		self._state = TVDBOffline(self)

	def __str__(self):

		return str(self._state)

	def __repr__(self):

		return "TVDB %s" % repr(self._state)

	def login(self, args):

		self._state.login(args)

	def lookup(self, text, args):

		return self._state.lookup(text, args.agent)

	def parse(self, show, db, args):

		return self._state.parse(show, db, args.agent)

	def change(self, state):

		self._state = state

	@staticmethod
	def accept(url):

		return url.isdigit()


class Epguides(object):

	def __init__(self):

		self.logger = logging.getLogger("Epguides")

	def __str__(self):

		return "epguides.com parser"

	def __repr__(self):

		return "Epguides()"

	@staticmethod
	def accept(url):

		return "epguides.com/" in url

	def login(self, args):

		pass

	def guess_encoding(self, response):

		raw = response.raw.read()
		text = raw.decode("iso-8859-1")

		if "charset=iso-8859-1" in text:
			return "iso-8859-1"

		return "utf8"

	def parse(self, show, db, args):

		headers = {"User-Agent": args.agent}
		response = requests.get(show.url, headers=headers)
		response.encoding = self.guess_encoding(response)

		for line in response.text.split("\n"):
			self._parse_line(line, show, db)

		show.updated = datetime.now()
		db.commit()

	def _parse_line(self, line, show, db):

		# Name of the show
		res = search(r"<title>(.*)</title>", line)
		if res:
			title = res.groups()[0]
			show.name = title.split(" (a ")[0]

		# Current status (running / ended)
		res = search(r'<span class="status">(.*)</span>', line)
		if res:
			text = res.groups()[0]
			if "current" in text:
				show.status = Show.RUNNING
			else:
				show.status = Show.ENDED
		else:
			if search(r"aired.*to.*[\d+]", line):
				show.status = Show.ENDED

		# Known formatting supported by this fine regex:
		# 4.     1-4            19 Jun 02  <a [..]>title</a>
		#   1.  19- 1   01-01    5 Jan 88  <a [..]>title</a>
		# 23     3-05           27/Mar/98  <a [..]>title</a>
		# 65.   17-10           23 Apr 05  <a [..]>title</a>
		# 101.   5-15           09 May 09  <a [..]>title</a>
		# 254.    - 5  05-254   15 Jan 92  <a [..]>title</a>
		res = search(r"^ *(\d+)\.? +(\d*)- ?(\d+) +([a-zA-Z0-9-]*)"\
		r" +(\d{1,2}[ /][A-Z][a-z]{2}[ /]\d{2}) *<a.*>(.*)</a>", line)

		if res:
			fields = res.groups()
			(total, season, epnum, prodnum, day, title) = fields

			day = day.replace("/", " ")
			airtime = datetime.strptime(day, "%d %b %y")

			self.logger.debug("Found episode %s", title)
			db.add_episode(Episode(title, season or 0, epnum,
					airtime.date(), prodnum, total), show)


class TVCom(object):

	def __init__(self):

		self._log = logging.getLogger("TVCom")

	def __str__(self):

		return "dummy tv.com parser to detect old urls"

	def __repr__(self):

		return "TVCom()"

	@staticmethod
	def accept(url):

		exp = "http://(www.)?tv.com/.*"
		return match(exp, url)

	def parse(self, source, _, __):

		self._log.error("The url %s is no longer supported", source.url)

	def login(self):

		pass
